module/VuFind/src/VuFind/Http/CachingDownloader.php

```diff
@@ -0,0 +1,168 @@
+
+/**
+ * Caching downloader.
+ *
+ * @category VuFind
+ * @package  Http
+ * @author   Mario Trojan <mario.trojan@uni-tuebingen.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class CachingDownloader
+{
+    /**
+     * CacheManager to update caches if necessary.
+     *
+     * @var CacheManager
+     */
+    protected $cacheManager;
+
+    /**
+     * Stored client options for cache key generation.
+     *
+     * @var array
+     */
+    protected $cacheOptions = [];
+
+    /**
+     * Cache to use for downloads
+     *
+     * @var StorageInterface
+     */
+    protected $cache;
+
+    /**
+     * HTTP service
+     *
+     * @var HttpService
+     */
+    protected $httpService;
+
+    /**
+     * Constructor
+     *
+     * @param HttpService $httpService  HTTP service
+     * @param string      $cacheManager Base directory for cache
+     */
+    public function __construct(HttpService $httpService, CacheManager $cacheManager)
+    {
+        $this->httpService = $httpService;
+        $this->cacheManager = $cacheManager;
+        $this->setCacheId('downloader');
+    }
+
+    /**
+     * Set a different cache.
+     *
+     * @param string $cacheId Cache ID
+     *
+     * @return void
+     */
+    public function setCacheId($cacheId)
+    {
+        $cacheName = $this->cacheManager->addDownloaderCache($cacheId);
+        $this->cache = $this->cacheManager->getCache($cacheName);
+    }
+
+    /**
+     * Download a resource using the cache in the background.
+     *
+     * @param string   $url              URL
+     * @param array    $params           Request parameters (e.g. additional headers)
+     * @param callable $validateCallback Callback for validation
+     *                                   before storing to cache
+     * @param callable $decodeCallback   Callback for decoding
+     *
+     * @return mixed
+     */
+    public function download(
+        $url,
+        $params=[],
+        callable $validateCallback=null,
+        callable $decodeCallback=null
+    ) {
+        $cacheItemKey = md5($url);
+        foreach ($params as $paramKey => $paramValue) {
+            $cacheItemKey .= '#' . $paramKey . '§' . $paramValue;
+        }
+
+        // Add new item to cache if not exists
+        if (!$this->cache->hasItem($cacheItemKey)) {
+            $response = $this->httpService->get($url, $params);
+            if (!$response->isOk()) {
+                throw new \Exception('Could not download URL: ' . $url);
+            }
+            $body = $response->getBody();
+            if ($validateCallback !== null && ($validateCallback($body) === false)) {
+                throw new \Exception('Invalid response body from URL: ' . $url);
+            }
+            $this->cache->addItem($cacheItemKey, $body);
+        }
+
+        $body = $this->cache->getItem($cacheItemKey);
+        if ($decodeCallback === null) {
```

Comment:
You could make this a little more concise with `return ($decodeCallback === null) ? $body : $decodeCallback($body);`

Reply:
I think the callbacks should have access to the full response, or at least status code and headers in addition to body. This would mean they need to be cached as well.